File: MyDailyBeat/API/VerveAPI/EVCUserSearchViewViewController.m
//
//  EVCSearchViewViewController.m
//  VerveAPI
//
//  Created by Virinchi Balabhadrapatruni on 11/2/14.
//  Copyright (c) 2014 eVerveCorp. All rights reserved.
//

#import "EVCUserSearchViewViewController.h"

@interface EVCUserSearchViewViewController ()

@end

@implementation EVCUserSearchViewViewController

- (id) init {
    
    self = [super initWithNibName:@"EVCUserSearchViewViewController_iPhone" bundle:[NSBundle bundleForClass:[self class]]];
    if (self) {
        
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    mSearchBar.delegate = self;
    mTableView.delegate = self;
    mTableView.dataSource = self;
    cancel = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel target:self action:@selector(cancelInvite)];
    self.navigationItem.leftBarButtonItem = cancel;
    self.navigationItem.leftItemsSupplementBackButton = YES;
    isFiltered = FALSE;
    mTableView.tableHeaderView = mSearchBar;
    [mSearchBar setShowsScopeBar:true];
    [mSearchBar setShowsCancelButton:false animated:true];
    [mSearchBar sizeToFit];
    self.navigationItem.title = @"Search for Users";
}

- (void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    VerveUser *user = [self.data objectAtIndex:indexPath.row];
    EVCUserInviteViewController *inviteWriter = [[EVCUserInviteViewController alloc] initWithGroup:self.groupToInviteTo andRecipient:user withSender:[[RestAPI getInstance] getCurrentUser]];
    [self.navigationController pushViewController:inviteWriter animated:YES];
}

- (void)cancelInvite {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void) searchBar:(UISearchBar *)searchBar selectedScopeButtonIndexDidChange:(NSInteger)selectedScope {
    switch (selectedScope) {
        case 0:
            type = SearchByScreenName;
            break;
        case 1:
            type = SearchByName;
            break;
        case 2:
            type = SearchByEmail;
            break;
            
    }
    
    NSString *text = searchBar.text;
    [self updateSearch:text];
}

- (void) searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText {
    if (searchText.length == 0 && isFiltered) {
        isFiltered = FALSE;
        [mTableView reloadData];
    }
    
}

- (void) updateSearch:(NSString *) text {
    dispatch_queue_t queue = dispatch_queue_create("dispatch_queue_t_dialog", NULL);
    dispatch_async(queue, ^{
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view makeToastActivity];
        });
        switch (type) {
            case SearchByScreenName:
                self.data = [currentSearch getUsersWithScreenNameContainingString:text withSortOrder:ASCENDING];
                break;
            case SearchByName:
                self.data = [currentSearch getUsersWithNameContainingString:text withSortOrder:ASCENDING];
                break;
            case SearchByEmail:
                self.data = [currentSearch getUsersWithEmailContainingString:text withSortOrder:ASCENDING];
                break;
        }
        [self removeUsersAlreadyInGroupFromSearchResults];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view hideToastActivity];
            [mTableView reloadData];
        });
        
    });
}

- (void) removeUsersAlreadyInGroupFromSearchResults {
    NSMutableArray *temp = [[NSMutableArray alloc] initWithArray:self.data copyItems:YES];
    self.data = [[NSMutableArray alloc] init];
    for (VerveUser * user in temp) {
        NSMutableArray *groupsForUser = [[RestAPI getInstance] getGroupsForUser:user];
        if (groupsForUser != nil) {
            BOOL member = NO;
            for (Group *g in groupsForUser) {
                if (g.groupID == self.groupToInviteTo.groupID) {
                    member = YES;
                }
            }
            if (!member) {
                [self.data addObject:user];
            }
        } else {
            [self.data addObject:user];
        }
    }
}

- (void)searchBarSearchButtonClicked:(UISearchBar *)searchBar
{
    [searchBar resignFirstResponder];
    NSString *text = [searchBar text];
    if(text.length == 0)
    {
        isFiltered = FALSE;
    }
    else
    {
        isFiltered = true;
        self.data = [[NSMutableArray alloc] init];
        currentSearch = [[EVCSearchEngine alloc] init];
        
        [self updateSearch:text];
        
    }
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    int rowCount;
    if(isFiltered)
        rowCount = (int) self.data.count;
    else
        rowCount = 1;
    
    return rowCount;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"Cell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil)
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:CellIdentifier];
    
    if (self.data.count == 0) {
        cell.textLabel.text = @"No Results Found";
        cell.imageView.image = nil;
        cell.accessoryType = UITableViewCellAccessoryNone;
        return cell;
    }
    
    if(isFiltered) {
        
        VerveUser *user = [self.data objectAtIndex:indexPath.row];
        cell.textLabel.text = user.screenName;
        dispatch_queue_t queue = dispatch_queue_create("dispatch_queue_t_dialog", NULL);
        dispatch_async(queue, ^{
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.view makeToastActivity];
            });
            NSURL *url = [[RestAPI getInstance] retrieveProfilePictureForUserWithScreenName:user.screenName];
            if (url != nil) {
                NSData *imageData = [NSData dataWithContentsOfURL:url];
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self.view hideToastActivity];
                    [cell.imageView setImage:[UIImage imageWithData:imageData]];
                });
            } else {
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self.view hideToastActivity];
                    [cell.imageView setImage:[UIImage imageNamed:@"default-avatar.png"]];
                });
            }
            
        });
        cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
        
        
        
    } else {
        cell.textLabel.text = @"No Results Found";
        cell.imageView.image = nil;
        cell.accessoryType = UITableViewCellAccessoryNone;
    }
    
    return cell;
}

@end

File: MDB_Swift/MyDailyBeat/API/src/EVCUserSearchViewViewController.swift
//  Converted with Swiftify v1.0.6242 - https://objectivec2swift.com/
//
//  EVCSearchViewViewController.h
//  VerveAPI
//
//  Created by Virinchi Balabhadrapatruni on 11/2/14.
//  Copyright (c) 2014 eVerveCorp. All rights reserved.
//
import UIKit
import Toast_Swift
protocol EVCUserSearchViewDelegate: NSObjectProtocol {
    func dismissUserSearch(_ controller: EVCUserSearchViewViewController)
}
public class EVCUserSearchViewViewController: UIViewController, UISearchBarDelegate, UITableViewDataSource, UITableViewDelegate {

    @IBOutlet var mTableView: UITableView!
    @IBOutlet var mSearchBar: UISearchBar!
    public var cancel: UIBarButtonItem!
    public var isFiltered: Bool = false
    public var type: UserSearchType = .searchByScreenName
    public var currentSearch: EVCSearchEngine!
    public var data = [VerveUser]()
    public var groupToInviteTo: Group!
    weak var delegate: EVCUserSearchViewDelegate?


    public init() {
        super.init(nibName: "EVCUserSearchViewViewController_iPhone", bundle: Bundle(for: EVCUserSearchViewViewController.self))
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }

    override public func viewDidLoad() {
        super.viewDidLoad()
        mSearchBar.delegate = self
        mTableView.delegate = self
        mTableView.dataSource = self
        cancel = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(self.cancelInvite))
        self.navigationItem.leftBarButtonItem = cancel
        self.navigationItem.leftItemsSupplementBackButton = true
        isFiltered = false
        mTableView.tableHeaderView = mSearchBar
        mSearchBar.showsScopeBar = true
        mSearchBar.setShowsCancelButton(false, animated: true)
        mSearchBar.sizeToFit()
        mSearchBar.autocorrectionType = .no
        mSearchBar.autocapitalizationType = .none
        self.navigationItem.title = "Search for Users"
    }

    public func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let user = self.data[indexPath.row]
        let inviteWriter = EVCUserInviteViewController()
        inviteWriter.groupToInviteTo = self.groupToInviteTo
        inviteWriter.recipient = user
        inviteWriter.sender = RestAPI.getInstance().getCurrentUser()
        self.navigationController?.pushViewController(inviteWriter, animated: true)
    }

    func cancelInvite() {
        self.dismiss(animated: true, completion: { _ in })
    }

    public func searchBar(_ searchBar: UISearchBar, selectedScopeButtonIndexDidChange selectedScope: Int) {
        switch selectedScope {
            case 0:
                type = .searchByScreenName
            case 1:
                type = .searchByName
            case 2:
                type = .searchByEmail
        default: break
        }

        let text: String = searchBar.text!
        self.updateSearch(text)
    }

    public func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
        if (searchText.characters.count ) == 0 && isFiltered {
            isFiltered = false
            mTableView.reloadData()
        }
    }

    func updateSearch(_ text: String) {
        
        DispatchQueue.global().async(execute: {() -> Void in
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.makeToastActivity(ToastPosition.center)
            })
            switch self.type {
                case .searchByScreenName:
                    self.data = self.currentSearch.getUsers(withScreenName: text, withSortOrder: .ascending)
                case .searchByName:
                    self.data = self.currentSearch.getUsers(withName: text, withSortOrder: .ascending)
                case .searchByEmail:
                    self.data = self.currentSearch.getUsers(withEmail: text, withSortOrder: .ascending)
            }

            self.removeUsersAlreadyInGroupFromSearchResults()
            DispatchQueue.main.async(execute: {() -> Void in
                self.view.hideToastActivity()
                self.mTableView.reloadData()
            })
        })
    }

    func removeUsersAlreadyInGroupFromSearchResults() {
        let temp: [VerveUser] = self.data
        self.data = [VerveUser]()
        for obj in temp {
            let user = obj
            let groupsForUser: [Group] = RestAPI.getInstance().getGroupsFor(user)
            var member: Bool = false
            let t = groupsForUser.map({ (group) -> Int in
                return group.groupID
            }).filter({ (id) -> Bool in
                return id == self.groupToInviteTo.groupID
            })
            if t.isEmpty {
                self.data.append(user)
            }
        }
    }

    public func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
        searchBar.resignFirstResponder()
        var text: String = searchBar.text!
        if (text.characters.count ) == 0 {
            isFiltered = false
        }
        else {
            isFiltered = true
            self.data = [VerveUser]()
            currentSearch = EVCSearchEngine()
            self.updateSearch(text)
        }
    }

    public func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        var rowCount: Int
        if isFiltered {
            rowCount = Int(self.data.count)
        }
        else {
            rowCount = 1
        }
        return rowCount
    }

    public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let CellIdentifier: String = "Cell"
        var cell: UITableViewCell? = tableView.dequeueReusableCell(withIdentifier: CellIdentifier)
        if cell == nil {
            cell = UITableViewCell(style: .subtitle, reuseIdentifier: CellIdentifier)
        }
        if self.data.count == 0 {
            cell?.textLabel?.text = "No Results Found"
            cell?.imageView?.image = nil
            cell?.accessoryType = .none
            return cell!
        }
        if isFiltered {
            let user: VerveUser = self.data[indexPath.row]
            cell?.textLabel?.text = user.screenName
            
            DispatchQueue.global().async(execute: {() -> Void in
                DispatchQueue.main.async(execute: {() -> Void in
                    self.view.makeToastActivity(ToastPosition.center)
                })
                let url: URL? = RestAPI.getInstance().retrieveProfilePictureForUser(withScreenName: (user.screenName))
                if url != nil {
                    let imageData: Data? = RestAPI.getInstance().fetchImage(atRemoteURL: url!)
                    DispatchQueue.main.async(execute: {() -> Void in
                        self.view.hideToastActivity()
                        cell?.imageView?.image = UIImage(data: imageData!)
                    })
                }
                else {
                    DispatchQueue.main.async(execute: {() -> Void in
                        self.view.hideToastActivity()
                        cell?.imageView?.image = UIImage(named: "default-avatar.png")
                    })
                }
            })
            cell?.accessoryType = .disclosureIndicator
        }
        else {
            cell?.textLabel?.text = "No Results Found"
            cell?.imageView?.image = nil
            cell?.accessoryType = .none
        }
        return cell!
    }
}
